### ISSUE TITLE
Major overuse of memory in HTTP/2 when transferring large files

### ISSUE THREAD

**Issue by bryancall** [apache/trafficserver#2530]
The overuse of buffers happens when we write to the response_buffer and then consume from the buf_reader in  Http2Stream::update_write_request().  The amount of buffer space in response_buffer is never checked and can grow to very large levels.  Another side effect is that it stalls the response to the client for 8 seconds while it is allocating the buffer block.

I tried a patch that would not consume from the buf_reader when the response_buffer was over 1MB for testing to see if this would work.  The result was that it stalled connection.

```
Client:
14:15:40 homer:/var/www/html$ nghttp -v -n https://127.0.0.1:4443/Large_file.m4v >& ~/out
19:41:48 homer:~$ less ~/out
[  8.394] recv (stream_id=13) :status: 200

Server:
14:15:46 homer:/opt$ sudo kill -SIGUSR1 $(pidof traffic_server); sleep 1 ; tail -100 /usr/local/var/log/trafficserver/traffic.out
...
         1341128704 |         1340997632 |      32768 | memory/ioBufAllocator[8]
...
     81857 |      40931 |           1341034496 |      32767 | memory/IOBuffer/HttpSM.cc:6648
```

**Comment by bryancall** [comment]
@masaori335 @maskit 
I was thinking about my next attempt to get this to work would be to look at HttpTunnel::consumer_reenable() and get high water marks working.  It looks like watermarks don't have any effect on this.  What are your thoughts?

**Comment by SolidWallOfCode** [comment]
The flow control mechanisms don't work for this? Can the HTTP/2 client session stall a single inbound stream in response to back up from the corresponding `HttpSM`?

**Comment by bryancall** [comment]
Yes, the flow control mechanisms never look to see how much buffer space is being used by the Http2Stream.  Yes, the Http2Steam could stall and restart the corresponding HttpSM.  I tried this by not consuming on the HttpTunnelProducer::read_buffer and then consuming when the Http2Stream buffer was below a hardcoded value.  But it seems like there needs to be something to trigger the HttpTunnel to restart reading from the origin connection.

**Comment by bryancall** [comment]
I am not seeing this problem on HTTP/1.1 with or without TLS.

**Comment by bryancall** [comment]
@SolidWallOfCode suggested sending a WRITE_READY event to the tunnel on the IRC.  I will try that out and see.
